server/src/main/java/io/crate/expression/scalar/ArrayUpperFunction.java

```diff
@@ -225,6 +226,20 @@ public Query toQuery(Function parent, Function arrayLength, Context context) {
         }
     }
 
+    private static Query docValueCountOrGeneric(Function parent,
+                                                 LuceneQueryBuilder.Context context,
+                                                 Reference arrayRef,
+                                                 IntPredicate valueCountIsMatch) {
+        BooleanQuery.Builder query = new BooleanQuery.Builder();
+        return query
+            .add(
+                NumTermsPerDocQuery.forArray(arrayRef, valueCountIsMatch),
+                BooleanClause.Occur.SHOULD
+            )
+            .add(genericFunctionFilter(parent, context), BooleanClause.Occur.MUST)
+            .build();
```

Comment:
I think both clauses would need to be `SHOULD` with a `setMinimumNumberShouldMatch(1);` - otherwise it will always have to run the genericFunctionFilter.